### ISSUE TITLE
Compiling for NGINX without Apache

### ISSUE THREAD

**Issue by infinitnet** [SpiderLabs/ModSecurity#686]
Even though I use --disable-apache2-module to compile the standalone module, configure fails with "configure: error: couldn't find APXS". Now to install APXS I have to install the Apache development packages. The problem is, that with Debian based systems, even if you use something like --no-install-recommends with apt, it forces you to install Apache and not only the development libraries. Is there any way around this? I have NGINX running and don't want to bloat my system with Apache. It would cause other issues too, because I'd always have to keep an eye on Apache after OS updates, so it doesn't start automatically and interferes with NGINX.

**Comment by msva** [comment]
@infinitnet can you reopen issue, since it is still impossible to build ModSecurity without apache headers and, probably, apsx too?

**Comment by msva** [comment]
thanks ;)

**Comment by zimmerle** [comment]
Already reported on #603. Tracking the progress on: #661.

@msva @infinitnet: Can we close this?

**Comment by msva** [comment]
Uhm... As you wish, of course, but it seems, there is no activity in the mentioned issues for a year and two years respectively ;)

// and then I don't get the purpose of nginx-refactoring branch. Isn't it for exactly that issue?

// Anyway, if you're sure it is enough to have mentioned issues to keep mind on possibility to apacheless install, you can freely close this issue ;)

**Comment by zimmerle** [comment]
By having #661 closed we will be able to have a nginx version free of Apache. Among of other things the nginx_refactoring is indeed built with that in mind. Although, it is not ready yet as you may have notice. Hope to have good news on that front soon. 

I will let this open. It is now marked as depending on #661.

**Comment by infinitnet** [comment]
@msva I re-opened the issue for you. Are you still unable to compile it without NGINX? Since I know how frustrating it can be, here's a script for you that I just wrote (it's similar to what I use currently) and that should work on Wheezy and Jessie (on Jessie only if you adjust the init file and update-rc.d commands, as that's now systemctl, but I guess that's not the important part about this anyway):

```
#!/bin/bash

nginx_dl=https://github.com/nginx/nginx/archive/master.zip
nginx_dir=nginx-master
ps_dl=https://github.com/pagespeed/ngx_pagespeed/archive/release-1.9.32.4-beta.zip
psol_dl=https://dl.google.com/dl/page-speed/psol/1.9.32.4.tar.gz
modsec_dl=https://github.com/SpiderLabs/ModSecurity/archive/nginx_refactoring.zip

apt-get -y --force-yes update
apt-get -y --force-yes install make automake patch gcc geoip-database openssl libssl-dev libpcre3 libpcre3-dev perl-modules libghc-zlib-dev libtool gettext wget curl build-essential zlib1g-dev libperl-dev libjemalloc-dev python-geoip libxml2 libxml2-dev libxml2-utils apache2-threaded-dev libcurl3-dev

mkdir -p /tmp/sources
wget -O /tmp/sources/$nginx_dir.zip $nginx_dl
wget -O /tmp/sources/pagespeed.zip $ps_dl
wget -O /tmp/sources/modsec.zip $modsec_dl
cd /tmp/sources/
unzip $nginx_dir.zip
unzip modsec.zip
unzip pagespeed.zip
rm -f $nginx_dir.zip
rm -f modsec.zip
rm -f pagespeed.zip
cd ngx_pagespeed*
wget $psol_dl
tar xfz "$ps_ver".tar.gz
rm -f "$ps_ver".tar.gz

useradd -M -d /var/www -s /sbin/nologin nginx
mkdir -p /etc/nginx/conf.d
mkdir -p /etc/nginx/conf.d/ssl
mkdir -p /usr/share/nginx
mkdir -p /var/log/nginx
mkdir -p /var/lib/nginx/tmp/client_body
mkdir -p /var/lib/nginx/tmp/proxy
mkdir -p /var/www
mkdir -p /var/cache/pagespeed
chown -R nginx:nginx /var/lib/nginx
chown -R nginx:nginx /var/log/nginx
chown -R nginx:nginx /var/cache/pagespeed

cd /tmp/sources/ModSecurity-nginx_refactoring
./autogen.sh
./configure --enable-standalone-module
make
make install
cp unicode.mapping /etc/nginx/conf.d/

cd /tmp/sources/$nginx_dir
./configure --user=nginx --group=nginx --prefix=/usr/share/nginx --sbin-path=/usr/sbin/nginx --conf-path=/etc/nginx/nginx.conf --error-log-path=/var/log/nginx/error.log --http-log-path=/var/log/nginx/access.log --http-client-body-temp-path=/var/lib/nginx/tmp/client_body --http-proxy-temp-path=/var/lib/nginx/tmp/proxy --http-fastcgi-temp-path=/var/lib/nginx/tmp/fastcgi --http-uwsgi-temp-path=/var/lib/nginx/tmp/uwsgi --http-scgi-temp-path=/var/lib/nginx/tmp/scgi --pid-path=/run/nginx.pid --lock-path=/var/lock/subsys/nginx --with-http_ssl_module --with-http_realip_module --with-http_sub_module --with-http_gzip_static_module --with-http_stub_status_module --with-http_perl_module --with-http_geoip_module --with-cc-opt='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -march=native -mtune=native' --add-module=/tmp/sources/ngx_pagespeed-release-1.9.32.4-beta --add-module=/tmp/sources/ModSecurity-nginx_refactoring/nginx/modsecurity
make
make install
/bin/cp -f /usr/share/GeoIP/GeoIP.dat /etc/nginx/

cat /etc/init.d/nginx << 'EOF'
#!/bin/sh
### BEGIN INIT INFO
# Provides:          nginx
# Required-Start:    $network $local_fs 
# Required-Stop:     $network $local_fs
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Stop/start nginx
### END INIT INFO

# Author: Sergey Budnevitch <sb@nginx.com>

PATH=/sbin:/usr/sbin:/bin:/usr/bin
DESC=nginx
NAME=nginx
CONFFILE=/etc/nginx/nginx.conf
DAEMON=/usr/sbin/nginx
PIDFILE=/var/run/$NAME.pid
SCRIPTNAME=/etc/init.d/$NAME

[ -x $DAEMON ] || exit 0

[ -r /etc/default/$NAME ] && . /etc/default/$NAME

DAEMON_ARGS="-c $CONFFILE $DAEMON_ARGS"

. /lib/init/vars.sh

. /lib/lsb/init-functions

do_start()
{
    start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON -- \
        $DAEMON_ARGS
    RETVAL="$?"
    return "$RETVAL"
}

do_stop()
{
    # Return
    #   0 if daemon has been stopped
    #   1 if daemon was already stopped
    #   2 if daemon could not be stopped
    #   other if a failure occurred
    start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --pidfile $PIDFILE --name $NAME
    RETVAL="$?"
    rm -f $PIDFILE
    return "$RETVAL"
}

do_reload() {
    #
    start-stop-daemon --stop --signal HUP --quiet --pidfile $PIDFILE --name $NAME
    RETVAL="$?"
    return "$RETVAL"
}

do_configtest() {
    if [ "$#" -ne 0 ]; then
        case "$1" in
            -q)
                FLAG=$1
                ;;
            *)
                ;;
        esac
        shift
    fi
    $DAEMON -t $FLAG -c $CONFFILE
    RETVAL="$?"
    return $RETVAL
}

do_upgrade() {
    OLDBINPIDFILE=$PIDFILE.oldbin

    do_configtest -q || return 6
    start-stop-daemon --stop --signal USR2 --quiet --pidfile $PIDFILE --name $NAME
    RETVAL="$?"
    sleep 1
    if [ -f $OLDBINPIDFILE -a -f $PIDFILE ]; then
        start-stop-daemon --stop --signal QUIT --quiet --pidfile $OLDBINPIDFILE --name $NAME
        RETVAL="$?"
    else
        echo $"Upgrade failed!"
        RETVAL=1
        return $RETVAL
    fi
}

case "$1" in
    start)
        [ "$VERBOSE" != no ] && log_daemon_msg "Starting $DESC " "$NAME"
        do_start
        case "$?" in
            0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
            2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
        esac
        ;;
    stop)
        [ "$VERBOSE" != no ] && log_daemon_msg "Stopping $DESC" "$NAME"
        do_stop
        case "$?" in
            0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
            2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
        esac
        ;;
  status)
        status_of_proc "$DAEMON" "$NAME" && exit 0 || exit $?
        ;;
  configtest)
        do_configtest
        ;;
  upgrade)
        do_upgrade
        ;;
  reload|force-reload)
        log_daemon_msg "Reloading $DESC" "$NAME"
        do_reload
        log_end_msg $?
        ;;
  restart|force-reload)
        log_daemon_msg "Restarting $DESC" "$NAME"
        do_configtest -q || exit $RETVAL
        do_stop
        case "$?" in
            0|1)
                do_start
                case "$?" in
                    0) log_end_msg 0 ;;
                    1) log_end_msg 1 ;; # Old process is still running
                    *) log_end_msg 1 ;; # Failed to start
                esac
                ;;
            *)
                # Failed to stop
                log_end_msg 1
                ;;
        esac
        ;;
    *)
        echo "Usage: $SCRIPTNAME {start|stop|status|restart|reload|force-reload|upgrade|configtest}" >&2
        exit 3
        ;;
esac

exit $RETVAL
EOF
chmod +x /etc/init.d/nginx
update-rc.d -f nginx defaults
/etc/init.d/nginx restart
```

**Comment by zimmerle** [comment]
@infinitnet yeah I know :( hopefully we can reduce the amount of all unnecessary dependencies soon.

Btw, that script seems great. I think it will be very valuable for the community in general. Want to start wiki page for nginx? Maybe place it on Gist? To avoid pollute this issue lets talk about it via email.

**Comment by msva** [comment]
@infinitnet
Main issue: I'm not using Ubuntu nor Debian 😉
Second issue: you're still installing a part of apache2: `apache2-threaded-dev` and I talking about entirely apacheless installation 😉


@zimmerle script is very ubuntu-centric (and don't take care on systemd-users), but, it probably be nice to have it somewhere just as cheatsheet for somebody ;)

**Comment by infinitnet** [comment]
@msva it's a quick and dirty workaround to at least not have the full Apache packages installed, it is obvious that completely Apache-less isn't possible yet. If that's not good enough for you, I guess you'll have to be patient.

@zimmerle the script is quite dirty and rather and example, but sure, if you think it would help people, I can start a wiki page on how to compile ModSecurity with NGINX when I have the time.

**Comment by wellumies** [comment]
I think it is pointless. If you are able to compile modsec with nginx you still have to turn off request processing to let it actually pass anything to the backend servers (read anything over 8k)

It is like using antivirus that does not detect any viruses, but uses 25% of your CPU

**Comment by victorhora** [comment]
Won't fix in 2.x. libModSecurity no longer depends on Apache or APR.

Further information available here - https://github.com/SpiderLabs/ModSecurity-nginx
